internal/database/database.go

```diff
@@ -933,6 +942,102 @@ WHERE (sources_secrets.ref_type = 'git_credentials' OR sources_secrets.ref_type
 	})
 }
 
+func (d *Database) GetSecret(ctx context.Context, principal string, name string) (*config.SecretRef, error) {
+	secrets, _, err := d.ListSecrets(ctx, principal, ListOptions{name: name})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(secrets) == 0 {
+		return nil, ErrNotFound
+	}
+
+	return secrets[0], nil
+}
+
+func (d *Database) DeleteSecret(ctx context.Context, principal string, name string) error {
+	return tx1(ctx, d, func(tx *sql.Tx) error {
+		if err := d.prepareDelete(ctx, tx, principal, "secrets", name, "secrets.manage"); err != nil {
+			return err
+		}
+		if err := d.delete(ctx, tx, "secrets", "name", name); err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+func (d *Database) ListSecrets(ctx context.Context, principal string, opts ListOptions) ([]*config.SecretRef, string, error) {
+	return tx3(ctx, d, func(txn *sql.Tx) ([]*config.SecretRef, string, error) {
+		expr, err := authz.Partial(ctx, authz.Access{
+			Principal:  principal,
+			Resource:   "secrets",
+			Permission: "secrets.view",
+		}, map[string]authz.ColumnRef{
+			"input.name": {Table: "secrets", Column: "name"},
+		})
+		if err != nil {
+			return nil, "", err
+		}
+
+		conditions, args := expr.SQL(d.arg, nil)
+		query := `SELECT
+        secrets.id,
+        secrets.name
+    FROM
+        secrets
+    WHERE (` + conditions + ")"
+
+		if opts.name != "" {
+			query += fmt.Sprintf(" AND (secrets.name = %s)", d.arg(len(args)))
+			args = append(args, opts.name)
+		}
+
+		if after := opts.cursor(); after > 0 {
+			query += fmt.Sprintf(" AND (secrets.id > %s)", d.arg(len(args)))
```

Comment:
[question]: Probably very out-of-scope for this PR, but... is there a way that we might be able to tap into [prepared statements](https://pkg.go.dev/database/sql#Tx.Stmt), instead of doing direct string substitution in the future? 🤔 

I'm fine creating a GH task issue for this, if we think it's something worth changing.

Reply:
The conditions come from PE and are created with each incoming request.... so it might be difficult? But, OTOH, the Rego never changes, so the number of possible statements is bounded, and we could therefore prepare them, couldn't we? (Probably has to be done lazily?)

Reply:
I guess my thought was that the string templating we're doing with `%s` could at least be replaced with something "stronger" in the future. The string chunks for the WHERE clause that's coming from PE should already be sanitized/safely-quoted, if memory serves.